File: apps/ios/Orange Cloud/Orange Cloud/ViewModels/DNSSettingsViewModel.swift
//
//  DNSSettingsViewModel.swift
//  Orange Cloud
//
//  Zone DNS 设置 + DNSSEC。两条链路权限不同（zone-dns-settings.* / dns.*），
//  任一读失败不连累另一条。
//

import Foundation
import Observation

@Observable
@MainActor
final class DNSSettingsViewModel {

    private(set) var settings: ZoneDNSSettings?
    private(set) var dnssec: ZoneDNSSEC?
    var isLoading = false
    var isMutating = false
    var error: String?
    var didMutate = false

    private let service: DNSSettingsService
    private let zoneId: String

    init(service: DNSSettingsService, zoneId: String) {
        self.service = service
        self.zoneId = zoneId
    }

    var settingsLoaded: Bool { settings != nil }
    var dnssecLoaded: Bool { dnssec != nil }

    func load() async {
        isLoading = true
        error = nil
        // 两条链路各自独立：DNSSEC 无权限时不该让 DNS 设置也不显示
        async let settingsTask = service.settings(zoneId: zoneId)
        async let dnssecTask = service.dnssec(zoneId: zoneId)
        settings = try? await settingsTask
        dnssec = try? await dnssecTask
        isLoading = false
    }

    private func apply(_ update: ZoneDNSSettingsUpdate) async {
        guard !isMutating else { return }
        isMutating = true
        error = nil
        do {
            settings = try await service.updateSettings(zoneId: zoneId, update: update)
            didMutate.toggle()
        } catch {
            self.error = error.localizedDescription
        }
        isMutating = false
    }

    func setFlattenAllCnames(_ on: Bool) async {
        await apply(ZoneDNSSettingsUpdate(flattenAllCnames: on))
    }

    func setMultiProvider(_ on: Bool) async {
        await apply(ZoneDNSSettingsUpdate(multiProvider: on))
    }

    func setSecondaryOverrides(_ on: Bool) async {
        await apply(ZoneDNSSettingsUpdate(secondaryOverrides: on))
    }

    func setZoneMode(_ mode: ZoneMode) async {
        await apply(ZoneDNSSettingsUpdate(zoneMode: mode.rawValue))
    }

    /// 高级名称服务器：用 nameservers.type 而非已弃用的 foundation_dns
    func setAdvancedNameservers(_ on: Bool) async {
        await apply(ZoneDNSSettingsUpdate(
            nameservers: ZoneNameserversUpdate(type: on ? "cloudflare.advanced" : "cloudflare.standard")
        ))
    }

    func setDNSSEC(_ on: Bool) async {
        guard !isMutating else { return }
        isMutating = true
        error = nil
        do {
            dnssec = try await service.setDNSSEC(zoneId: zoneId, enabled: on)
            didMutate.toggle()
        } catch {
            self.error = error.localizedDescription
        }
        isMutating = false
    }
}

File: apps/android/app/src/main/kotlin/jiamin/chen/orangecloud/ui/dnssettings/DnsSettingsViewModel.kt
package jiamin.chen.orangecloud.ui.dnssettings

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import jiamin.chen.orangecloud.core.auth.AuthRepository
import jiamin.chen.orangecloud.core.auth.Scopes
import jiamin.chen.orangecloud.data.model.ZoneDnsSettings
import jiamin.chen.orangecloud.data.model.ZoneDnsSettingsUpdate
import jiamin.chen.orangecloud.data.model.ZoneDnssec
import jiamin.chen.orangecloud.data.model.ZoneNameserversUpdate
import jiamin.chen.orangecloud.data.repository.DnsSettingsRepository
import kotlinx.coroutines.async
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.receiveAsFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

data class DnsSettingsUiState(
    val zoneName: String = "",
    val settings: ZoneDnsSettings? = null,
    val dnssec: ZoneDnssec? = null,
    val isLoading: Boolean = false,
    val isMutating: Boolean = false,
    val canWriteSettings: Boolean = false,
    val canWriteDns: Boolean = false,
)

sealed interface DnsSettingsEvent {
    data class Error(val message: String?) : DnsSettingsEvent
}

@HiltViewModel
class DnsSettingsViewModel @Inject constructor(
    savedStateHandle: SavedStateHandle,
    private val repository: DnsSettingsRepository,
    authRepository: AuthRepository,
) : ViewModel() {

    private val zoneId: String = checkNotNull(savedStateHandle["zoneId"])
    private val canWriteSettings = authRepository.hasScope(Scopes.ZONE_DNS_SETTINGS_WRITE)
    private val canWriteDns = authRepository.hasScope(Scopes.DNS_WRITE)

    private val _uiState = MutableStateFlow(
        DnsSettingsUiState(
            zoneName = savedStateHandle.get<String>("zoneName").orEmpty(),
            canWriteSettings = canWriteSettings,
            canWriteDns = canWriteDns,
        ),
    )
    val uiState: StateFlow<DnsSettingsUiState> = _uiState.asStateFlow()

    private val eventChannel = Channel<DnsSettingsEvent>(Channel.BUFFERED)
    val events: Flow<DnsSettingsEvent> = eventChannel.receiveAsFlow()

    init { load() }

    fun load() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true) }
            // 两条链路权限不同，任一失败不连累另一条
            val s = async { runCatching { repository.settings(zoneId) }.getOrNull() }
            val d = async { runCatching { repository.dnssec(zoneId) }.getOrNull() }
            val settings = s.await()
            val dnssec = d.await()
            _uiState.update { it.copy(settings = settings, dnssec = dnssec, isLoading = false) }
        }
    }

    private fun apply(update: ZoneDnsSettingsUpdate) {
        if (!canWriteSettings || _uiState.value.isMutating) return
        viewModelScope.launch {
            _uiState.update { it.copy(isMutating = true) }
            runCatching { repository.updateSettings(zoneId, update) }
                .onSuccess { s -> _uiState.update { it.copy(settings = s) } }
                .onFailure { eventChannel.send(DnsSettingsEvent.Error(it.message)); load() }
            _uiState.update { it.copy(isMutating = false) }
        }
    }

    fun setFlattenAllCnames(on: Boolean) = apply(ZoneDnsSettingsUpdate(flattenAllCnames = on))

    fun setMultiProvider(on: Boolean) = apply(ZoneDnsSettingsUpdate(multiProvider = on))

    fun setZoneMode(mode: String) = apply(ZoneDnsSettingsUpdate(zoneMode = mode))

    /** 高级名称服务器：用 nameservers.type 而非已弃用的 foundation_dns */
    fun setAdvancedNameservers(on: Boolean) = apply(
        ZoneDnsSettingsUpdate(
            nameservers = ZoneNameserversUpdate(if (on) "cloudflare.advanced" else "cloudflare.standard"),
        ),
    )

    fun setDnssec(on: Boolean) {
        if (!canWriteDns || _uiState.value.isMutating) return
        viewModelScope.launch {
            _uiState.update { it.copy(isMutating = true) }
            runCatching { repository.setDnssec(zoneId, on) }
                .onSuccess { d -> _uiState.update { it.copy(dnssec = d) } }
                .onFailure { eventChannel.send(DnsSettingsEvent.Error(it.message)); load() }
            _uiState.update { it.copy(isMutating = false) }
        }
    }
}
